IntegrationTests/tests_01_allocation_counters/test_01_resources/shared.swift

```diff
@@ -0,0 +1,49 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020-2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import NIOEmbedded
+import NIOHTTP2
+
+/// Test use only. Allows abstracting over the two multiplexer implementations to write common testing code
+internal protocol MultiplexerChannelCreator {
+    func createStreamChannel(promise: EventLoopPromise<Channel>?, _ streamStateInitializer: @escaping NIOHTTP2Handler.StreamInitializer)
+}
+
+extension HTTP2StreamMultiplexer: MultiplexerChannelCreator { }
+extension NIOHTTP2Handler.StreamMultiplexer: MultiplexerChannelCreator { }
+
+/// Have two `EmbeddedChannel` objects send and receive data from each other until
+/// they make no forward progress.
+func interactInMemory(_ first: EmbeddedChannel, _ second: EmbeddedChannel) throws {
+    var operated: Bool
+
+    func readBytesFromChannel(_ channel: EmbeddedChannel) throws -> ByteBuffer? {
+        return try channel.readOutbound(as: ByteBuffer.self)
+    }
+
+    repeat {
+        operated = false
+        first.embeddedEventLoop.run()
```

Comment:
The correctness of this is predicated on the two channels having the same EL: can you precondition on this in the entrance to the function?